src/main/kotlin/org/rust/lang/core/types/infer/Expressions.kt

```diff
@@ -40,28 +40,38 @@ fun inferExpressionType(expr: RsExpr) = when (expr) {
 }
 
 private fun inferPathExprType(expr: RsPathExpr): Ty {
-    val target = expr.path.reference.resolve() as? RsNamedElement
-        ?: return TyUnknown
+    val boundElement = expr.path.reference.advancedResolve() ?: return TyUnknown
+    val element = boundElement.element as? RsNamedElement ?: return TyUnknown
 
-    return inferDeclarationType(target)
+    return inferDeclarationType(element).substitute(boundElement.typeArguments)
 }
 
 private fun inferStructLiteralType(expr: RsStructLiteral): Ty {
-    val base = expr.path.reference.resolve()
-    return when (base) {
-        is RsStructItem -> inferStructTypeParameters(expr, base)
-        is RsEnumVariant -> inferEnumTypeParameters(expr, base)
+    val boundElement = expr.path.reference.advancedResolve()
+    val element = boundElement?.element
+    return when (element) {
+        is RsStructItem -> element.type
+            .substitute(boundElement.typeArguments)
+            .substitute(inferStructTypeArguments(expr))
+        is RsEnumVariant -> element.parentEnum.type
+            .substitute(boundElement.typeArguments)
+            .substitute(inferStructTypeArguments(expr))
         else -> TyUnknown
     }
 }
 
 private fun inferCallExprType(expr: RsCallExpr): Ty {
     val fn = expr.expr
     if (fn is RsPathExpr) {
-        val variant = fn.path.reference.resolve()
-        when (variant) {
-            is RsEnumVariant -> return inferTupleEnumTypeParameters(expr, variant)
-            is RsStructItem -> return inferTupleStructTypeParameters(expr, variant)
+        val boundElement = fn.path.reference.advancedResolve()
```

Comment:
Hm, I was going to say that we should use `val (elemet, subst) = fn.path.refenrece.advancedResolve()` but looks like it's not going to work because of nulls. 

Still, let's make it a convention to destruct `BoundElement`s, and to use `subst` for the substitution. (And yeah, should probably rename `TypeArguments` to `Substitution`).